R/utils.R

```diff
@@ -202,25 +210,357 @@ find_db_tablename <- function(dbcon, drm_table, verbose = TRUE) {
+                        interval = NULL, # for numeric inputs only
+                        colnames = NULL, # for data.table/.frame inputs only
+                        coltypes = NULL, #          "-"
+                        unique = FALSE) { #          "-"
+
+
+  ## Get argument names and restructure input
+  if (any(class(arginput) == "list")) {
+    # Note: Users can provide multiple arginputs to be checked by combining them
+    # into a list ...or they might want to check an arginput that is supposed to
+    # be a list itself
+    # Here: We infer which one it is based on deparse(substitute)
+    # If arginput is provided as a single input name:
+    # -> assume input itself is supposed to be a list
+    # If each list item corresponds to a separate argument name
+    # -> assume user wants to check individual items
+    # it's a bit hacky but seems to work for tested scenarios
+    argnames <- sapply(substitute(arginput), deparse)[-1]
+    if (length(argnames) < 1) {
+      argnames <- deparse(substitute(arginput))
+      arginput <- list(arginput = arginput)
+    }
+  } else {
+    # get name of argument
+    argnames <- deparse(substitute(arginput))
+
+    # turn arginput into list (for Map function below to work)
+    arginput <- list(arginput = arginput)
+  }
+
+
+  ## Define new function to check for integers
+  #  Note: base R's `is.integer` does not return TRUE if type == numeric
+  #  Note 2: For coltypes check below, this function is not used
+  #  (instead coltypes are checked for whether class(column) returns "integer")
+  is_integer <- function(x) {
+    if (is.numeric(x)) {
+      tol <- .Machine$double.eps
+      return(abs(x - round(x)) < tol)
+    } else {
+      return(FALSE)
+    }
+  }
+
+
+  ## Function defining all input checks
+  run_checks <- function(arginput, argname) {
+    ###### CHECK 1 (for all input types): Check if type is correct
+    ## For DB connections
+    if (any(grepl("dbi|con|posgre|sql", argtype, ignore.case = TRUE))) {
+      if (!RPostgreSQL::isPostgresqlIdCurrent(arginput) &&
+        !grepl("PostgreSQL", class(arginput)[1])) {
+        stop(
+          paste0(
+            "Invalid user input in '",
+            as.character(sys.calls()[[1]])[1], "': '",
+            argname, "' needs to be a valid database connection.\n",
+            "We recommend the following method to establish a DB connection:\n",
+            "drv <- dbDriver('PostgreSQL')\n",
+            "dbcon <- DBI::dbConnect(drv, dbname = 'db_name', ",
+            "host = 'domain_name.net', port = 1234, ",
+            "user = getPass('Enter user:'), password = getPass('password'))\n",
+            "\nPlease refer to the function documentation for more details."
+          ),
+          call. = FALSE
+        )
+      }
+
+      ## For all other inputs
+    } else if ((any(argtype == "integer") && !all(is_integer(arginput))) ||
+      !any(argtype == "integer") && !any(class(arginput) %in% argtype)) {
```

Comment:
The check cannot go through in the following cases:
1. input is integer, argtype is "numeric";
2. input is float, argtype is c("numeric", "integer")

code to reproduce the result:
`test <- function(a, ...){
  check_input(a, ...)
}
test(c(1L, 2L), argtype = "numeric")
test(c(1.5, 2.5), argtype = c("numeric", "integer"))
`

Reply:
I think maybe prioritize resolve case 1 as it is more realistic and both `integer` and `double` should be  `numeric`. Case 2 is much less likely to happen.

Reply:
1. Yes, you are right! Although `class(1L) == "numeric"` returns `FALSE`, for the purpose of this functions, it's probably a good idea to let class integer pass the numeric test. I just made a change adding another condition in the `if` statement to avoid returning an error when the input is an integer and argtype = "numeric" and also added a test scenario for that in the test file. I also added a note about this in the documentation (including a note about `coltypes` where the logic is slightly different). 
2. Right, I think the best way to handle this is to add an explanation in the documentation. Since integer is a special case of numeric, I think checking whether the input is either or is not very meaningful. Let me know if this makes sense.